# Import libraries

In [6]:
import os
import pandas as pd
import numpy as np

# Loading the dataset

In [7]:
def get_data(directory_name:str, file_name:str):
    path = os.path.join(directory_name,f"{file_name}.csv")
    return pd.read_csv(path)

In [8]:
flat_data=get_data("/home/spynom/project/real_state_price_prediction_project/data/raw","scrapped_data")

# Preliminary analysis

In [9]:
flat_data.sample(10)

link             name  \
7388  https://www.makaan.com/noida/laureate-buildwel...  4 BHK Apartment   
5473  https://www.makaan.com/noida/builder-project-i...  3 BHK Apartment   
57    https://www.makaan.com/noida/ats-green-pious-h...  3 BHK Apartment   
3394  https://www.makaan.com/noida/supertech-ecociti...  2 BHK Apartment   
176   https://www.makaan.com/noida/gaursons-hi-tech-...  3 BHK Apartment   
2432  https://www.makaan.com/noida/kk-buildcom-krish...  3 BHK Apartment   
6054  https://www.makaan.com/noida/logix-group-bloss...  5 BHK Apartment   
3459  https://www.makaan.com/noida/supertech-emerald...              NaN   
6726  https://www.makaan.com/noida/fusion-the-rivule...  3 BHK Apartment   
7191  https://www.makaan.com/noida/godrej-properties...  3 BHK Apartment   

             size                                            address  \
7388  4,874 sq ft            Laureate Parx LaureateSector 108, Noida   
5473  1,600 sq ft                                  Sector 107, Noida   
57    1,675 sq ft               ATS Pious HideawaysSector 150, Noida   
3394    890 sq ft                 Supertech EcocitiSector 137, Noida   
176   1,450 sq ft  Gaursons 16th Park ViewSector 19 Yamuna Expres...   
2432  1,450 sq ft             KK Krishna Elite Homesnoida ext, Noida   
6054  4,500 sq ft              Logix Blossom CountySector 137, Noida   
3459          NaN                                                NaN   
6726  1,380 sq ft         Fusion The Rivulet Phase 1noida ext, Noida   
7191  1,249 sq ft                Godrej Godrej WoodsSector 43, Noida   

          price                                               rate  \
7388  6.31CrEMI   12,955/ sq ft the price/ sq ft of this proper...   
5473     74LEMI   4,625/ sq ft the price/ sq ft of this propert...   
57    1.75CrEMI   10,447/ sq ft the price/ sq ft of this proper...   
3394     70LEMI   7,865/ sq ft the price/ sq ft of this propert...   
176      92LEMI   6,344/ sq ft the price/ sq ft of this propert...   
2432   41.5LEMI   2,862/ sq ft the price/ sq ft of this propert...   
6054  3.75CrEMI   8,333/ sq ft the price/ sq ft of this propert...   
3459  22.35LEMI   3,519/ sq ft the price/ sq ft of this propert...   
6726  1.25CrEMI   9,058/ sq ft the price/ sq ft of this propert...   
7191  3.35CrEMI   26,851/ sq ft the price/ sq ft of this proper...   

     Carpet area              Status  Bathrooms New/Resale  ...  \
7388  4200 sq ft       Ready to move        3.0        New  ...   
5473         NaN  Under Construction        3.0        New  ...   
57    1430 sq ft       Ready to move        3.0     Resale  ...   
3394         NaN       Ready to move        2.0     Resale  ...   
176   1100 sq ft       Ready to move        3.0     Resale  ...   
2432  1300 sq ft       Ready to move        2.0        New  ...   
6054  4000 sq ft       Ready to move        5.0     Resale  ...   
3459         NaN                 NaN        NaN        NaN  ...   
6726   696 sq ft  Under Construction        2.0        New  ...   
7191         NaN  Under Construction        NaN        New  ...   

                                                RERA ID Age of Property.1  \
7388                                      UPRERAPRJ5998       6 - 7 years   
5473                                                NaN               NaN   
57                                      UPRERAPRJ442430               NaN   
3394                                      UPRERAPRJ9744       6 - 7 years   
176                                       UPRERAPRJ6801        0 - 1 year   
2432                                Rera Not Applicable        0 - 1 year   
6054                       UPRERAPRJ4466, UPRERAPRJ4411       6 - 7 years   
3459                                                NaN               NaN   
6726                                    UPRERAPRJ145736               NaN   
7191  UPRERAPRJ773536, UPRERAPRJ145318, UPRERAPRJ704730               NaN   

     Additional Rooms.1 Balconies.1  New/Resale.1 Ownership Type

In [10]:
flat_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9132 entries, 0 to 9131
Data columns (total 29 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   link                9132 non-null   object 
 1   name                9086 non-null   object 
 2   size                9086 non-null   object 
 3   address             9086 non-null   object 
 4   price               9032 non-null   object 
 5   rate                9032 non-null   object 
 6   Carpet area         4697 non-null   object 
 7   Status              9086 non-null   object 
 8   Bathrooms           7274 non-null   float64
 9   New/Resale          7799 non-null   object 
 10  Price Negotiable    1002 non-null   object 
 11  Age of Property     806 non-null    object 
 12  Floor               855 non-null    object 
 13  Additional Rooms    313 non-null    object 
 14  Balconies           123 non-null    float64
 15  Price Negotiable.1  8084 non-null   object 
 16  Securi

In [11]:
flat_data.shape

(9132, 29)

In [12]:
flat_data.dtypes

link                   object
name                   object
size                   object
address                object
price                  object
rate                   object
Carpet area            object
Status                 object
Bathrooms             float64
New/Resale             object
Price Negotiable       object
Age of Property        object
Floor                  object
Additional Rooms       object
Balconies             float64
Price Negotiable.1     object
Security Deposit       object
Facing                 object
Status.1               object
RERA ID                object
Age of Property.1      object
Additional Rooms.1     object
Balconies.1           float64
New/Resale.1           object
Ownership Type         object
Booking Amount         object
Overlooking            object
1                      object
2                      object
dtype: object

In [13]:
flat_data.duplicated().sum()

np.int64(1001)

**Observations:**
- There are 9132 instances and 29 attributes
- some attributes are repeated
- so many attributes having misssing values.
- most of attributes have inappropriate data types.
- total 1001 instances are duplicated

# Detail analysis

In [14]:
flat_data.columns

Index(['link', 'name', 'size', 'address', 'price', 'rate', 'Carpet area',
       'Status', 'Bathrooms', 'New/Resale', 'Price Negotiable',
       'Age of Property', 'Floor', 'Additional Rooms', 'Balconies',
       'Price Negotiable.1', 'Security Deposit', 'Facing', 'Status.1',
       'RERA ID', 'Age of Property.1', 'Additional Rooms.1', 'Balconies.1',
       'New/Resale.1', 'Ownership Type', 'Booking Amount', 'Overlooking', '1',
       '2'],
      dtype='object')

### name

In [15]:
flat_data["name"].unique()

array(['1 BHK Independent Floor', '3 BHK Apartment', '2 BHK Apartment',
       '1 BHK Apartment', '4 BHK Apartment', '5 BHK Apartment',
       '1 RK Studio Apartment', '3 BHK Independent Floor',
       '7 BHK Independent Floor', '4 BHK Independent Floor',
       '2 BHK Independent Floor', nan, '7 BHK Apartment',
       '6 BHK Apartment', '5 BHK Independent Floor'], dtype=object)

### size

In [16]:
flat_data["size"].sample(5)

3348    1,425 sq ft
2013    1,470 sq ft
2646    1,515 sq ft
4002    1,600 sq ft
4048      975 sq ft
Name: size, dtype: object

In [17]:
flat_data["size"].str.split(" ",n=1).str.get(1).unique()

array(['sq ft', nan], dtype=object)

### address

In [40]:
(flat_data.address.str.lower().str.replace(r'[A-Za-z0-9\s]*?4 greater noida wes+[A-Za-z0-9\s]*',"4 greater noida west",regex=True)
                    .str.replace(r'[A-Za-z0-9\s]*?(extension|ext)[A-Za-z0-9\s]*',"extension, noida",regex=True)
                    .str.replace(r'[A-Za-z0-9\s]*?ambedkar city[A-Za-z0-9\s]*',"ambedkar city, noida",regex=True)
                    .str.replace(r'[A-Za-z0-9\s]*?yamuna express[A-Za-z0-9\s]*',"yamuna expressway, noida",regex=True)
                    .str.replace("surajpur site 4, noida","surajpur site 4, greater noida")
                    .str.replace("knowledge park 3, noida","knowledge park 3, greater noida")
                    .str.replace("sikandarpur village, noida","sector 87, noida")
                    .str.replace("dallupura, noida","sector 10, noida")
                    .str.replace("pristine avenuegaur city road, noida","gaur city, greater noida")
                    .str.replace("migsun vilaasaeta ii, noida","s block, greater noida")
                    .str.replace("nirala greenshiregreater noida west road, noida","greater noida west road, noida")
                    .str.replace("home and soul f premiereyeida, noida","jal vayu vihar, noida")
                    .str.replace("apte gra indraprastha phase 2shramik kunj, noida","sector 93, noida")
                    .str.replace("apte gra indraprastha phase 2shramik kunj, noida","sector 93, noida")
                    .str.replace(r'[0-9a-z ]+sector',"sector" ,regex=True).str.replace("-",'')
                    .str.replace("noida, ",'')   
                    .str.replace(r"sector[\s0]*", "sector ",regex=True)
                    .str.replace(r"[A-Za-z0-9\s,]*greater noida[A-Za-z0-9\s,]*","greater-noida",regex=True)
                    .str.replace(r"[,]* noida",", noida",regex=True)
                    .str.replace('dream house sr construction dream house sector 73, noida',"sector 73, noida")
                    .str.replace(r"sector 93[\sa-z]+, noida","sector 93, noida",regex=True)
                    .replace("sector 143b, noida","sector 143, noida")
                    .replace('sector 16b, noida','sector 16, noida')
                    .replace('sector 140a, noida','sector 140, noida')
                    .replace('sector 15a, noida','sector 15, noida')
                    .replace('sector 25a, noida','sector 25, noida')
                    .replace("greater-noida", "greater noida")
                    
                        ).unique()

array(['sector 70, noida', 'sector 75, noida', 'yamuna expressway, noida',
       'sector 49, noida', 'sector 78, noida', 'sector 143, noida',
       'sector 74, noida', 'sector 93, noida', 'sector 53, noida',
       'sector 134, noida', 'sector 126, noida', 'sector 45, noida',
       'sector 151, noida', 'extension, noida', 'sector 137, noida',
       'sector 44, noida', 'sector 129, noida', 'sector 107, noida',
       'sector 168, noida', 'sector 16, noida', 'sector 128, noida',
       'sector 73, noida', 'sector 144, noida', 'sector 150, noida',
       'sector 82, noida', 'sector 100, noida', 'sector 72, noida',
       'sector 120, noida', 'sector 50, noida', 'sector 117, noida',
       'sector 108, noida', 'sector 79, noida', 'sector 77, noida',
       'sector 130, noida', 'sector 146, noida', 'sector 76, noida',
       'sector 121, noida', 'sector 104, noida', 'sector 101, noida',
       'sector 153, noida', 'sector 36, noida', 'greater noida',
       'sector 102, noida', 'sector 

In [13]:
temp=flat_data.copy()

In [14]:
temp["address"].fillna("sector").loc[lambda x:~ x.str.lower().str.contains("sector")].loc[lambda x: x.str.lower().str.contains("extension|ext",regex=True)].unique()

array(['Thv Rudransh Green Vistanoida ext, Noida',
       'Fusion The Rivulet Phase 1noida ext, Noida',
       'Thv AS Ultima Onenoida ext, Noida',
       'Gaursons India Gaur City 2 16th AvenueUrbainia Trinity Noida Extension Yakubpur Noida, Noida',
       'Golfgreen Elite XNoida Extensions, Noida',
       'Sublime Spring ElmasPhase 2 Noida Extension, Noida',
       'Thv Orchid Heightsnoida ext, Noida',
       'City Antalya HeightsNoida Extensions, Noida',
       'Path View NCR Lotus Tower Garden CityNoida Extensions, Noida',
       'Trinity Primusnoida ext, Noida',
       'Srijee Saraswati Smart HomesNoida Extensions, Noida',
       'Phase 2 Noida Extension, Noida',
       'Ambesten Twin Chaletnoida ext, Noida',
       'Trinity High Street South Avenuenoida ext, Noida',
       'Trinity High Street North Avenuenoida ext, Noida',
       'KK Krishna Elite Homesnoida ext, Noida',
       'Gaurik Krishna Homes 2Noida Extensions, Noida',
       'Sam Citynoida ext, Noida',
       'Ajay Eros 

In [15]:
flat_data["address"].str.split("Sector").str.get(1).str.replace("-","").str.strip().str.split(",").str.get(0).replace("Nodia","").str.strip().unique().size

120

### price

In [16]:
flat_data["price"].str.replace("EMI","").str.replace(r"[0-9.]","",regex=True).unique()

array(['L', 'Cr', nan, 'CrNegotiable', 'LNegotiable'], dtype=object)

In [17]:
df_=flat_data.assign(**{"amount":flat_data["price"].str.replace("EMI","").str.extract(r'(\d+\.\d+)').astype(float),
"unit":flat_data["price"].str.replace("EMI","").str.replace(r"[0-9.]","",regex=True)})

In [18]:
df_[(df_["amount"]>10)&(df_["unit"]=="Cr")&["name"]=="3 BHK Apartment"]

/tmp/ipykernel_372/3294693015.py:1: FutureWarning: Logical ops (and, or, xor) between Pandas objects and dtype-less sequences (e.g. list, tuple) are deprecated and will raise in a future version. Wrap the object in a Series, Index, or np.array before operating instead.
  df_[(df_["amount"]>10)&(df_["unit"]=="Cr")&["name"]=="3 BHK Apartment"]


Empty DataFrame
Columns: [link, name, size, address, price, rate, Carpet area, Status, Bathrooms, New/Resale, Price Negotiable, Age of Property, Floor, Additional Rooms, Balconies, Price Negotiable.1, Security Deposit, Facing, Status.1, RERA ID, Age of Property.1, Additional Rooms.1, Balconies.1, New/Resale.1, Ownership Type, Booking Amount, Overlooking, 1, 2, amount, unit]
Index: []

[0 rows x 31 columns]

### rate

In [19]:
flat_data["rate"].str.extract(r"([a-z]+ [a-z]+)").iloc[:,0].unique()

array(['sq ft', nan], dtype=object)

### Carpet area

In [20]:
flat_data["Carpet area"].str.extract(r"([a-z]+ [a-z]+)").iloc[:,0].unique()

array(['sq ft', nan], dtype=object)

### Status & Status.1

In [21]:
flat_data[["Status","Status.1"]].groupby(["Status","Status.1"]).count()

Empty DataFrame
Columns: []
Index: [(Ready to move, Furnished), (Ready to move, Semi-Furnished), (Ready to move, Unfurnished), (Under Construction, Furnished), (Under Construction, Semi-Furnished), (Under Construction, Unfurnished)]

### Bathrooms

In [22]:
flat_data["Bathrooms"].unique()

array([ 1.,  3.,  2.,  4.,  5., nan,  8.,  6.,  7.])

In [23]:
flat_data["Bathrooms"].dtype

dtype('float64')

### New/Resale & New/Resale.1

In [24]:
flat_data["New/Resale"].unique()

array(['Resale', 'New', nan], dtype=object)

In [79]:
flat_data[["New/Resale.1","New/Resale"]][~(flat_data["New/Resale.1"].isnull())&(flat_data["New/Resale.1"].isnull())]

Empty DataFrame
Columns: [New/Resale.1, New/Resale]
Index: []

### Price Negotiable & Price Negotiable.1

In [25]:
flat_data[["Price Negotiable","Price Negotiable.1"]][~(flat_data["Price Negotiable"].isnull())&~(flat_data["Price Negotiable.1"].isnull())]

Empty DataFrame
Columns: [Price Negotiable, Price Negotiable.1]
Index: []

In [26]:
(flat_data["Price Negotiable"].fillna("")+flat_data["Price Negotiable.1"].fillna("")).unique()

array(['No', '', 'Yes'], dtype=object)

In [27]:
flat_data["Price Negotiable"].unique()

array([nan, 'No', 'Yes'], dtype=object)

### Age of Property & Age of Property.1

In [28]:
(flat_data["Age of Property"].fillna("")+flat_data["Age of Property.1"].fillna("")).str.extract(r"([0-9]+ y)").loc[:,0].str.replace(" y","").unique()

array([nan, '5', '9', '3', '12', '6', '1', '8', '10', '2', '7', '4', '17',
       '15', '19', '14', '11', '21', '16', '13', '18', '22', '23', '25',
       '125'], dtype=object)

In [29]:
flat_data["Age of Property"].str.extract(r"([0-9]+ y)").loc[:,0].str.replace(" y","").unique()

array([nan, '1', '10', '9', '3', '2', '6', '7', '5', '11', '8'],
      dtype=object)

In [30]:
flat_data[["Age of Property.1","Age of Property"]][~(flat_data["Age of Property"].isnull())&~(flat_data["Age of Property.1"].isnull())]

Empty DataFrame
Columns: [Age of Property.1, Age of Property]
Index: []

In [31]:
flat_data["Age of Property"].unique()

array([nan, '0 - 1 year', '9 - 10 years', '8 - 9 years', '2 - 3 years',
       '2 years', '5 - 6 years', '6 - 7 years', '1 year', '1 - 2 years',
       '4 - 5 years', '10 - 11 years', '9 years', '7 - 8 years',
       '6 years', '10 years'], dtype=object)

In [32]:
flat_data["Age of Property"].str.extract(r"(\d+)").iloc[:,0].unique()

array([nan, '0', '9', '8', '2', '5', '6', '1', '4', '10', '7'],
      dtype=object)

### Floor

In [33]:
flat_data["Floor"].str.replace("Gr","0").str.extract(r"(\d+)").iloc[:,0].unique()

array([nan, '25', '22', '12', '16', '15', '17', '4', '6', '28', '1', '20',
       '30', '14', '5', '18', '10', '8', '2', '24', '0', '11', '23', '3',
       '21', '7', '9', '13', '31', '45', '27'], dtype=object)

In [34]:
flat_data["Floor"].unique()

array([nan, '25th of 32', '22nd of 32', '12th of 31', '16th of 31',
       '15th of 26', '25th of 26', '17th of 26', '4th of 26', '6th of 26',
       '12th of 26', '28th of 34', '6th of 34', '1st of 6', '25th of 47',
       '20th of 28', '1st of 17', '30th of 34', '20th of 47',
       '14th of 29', '12th of 28', '22nd of 29', '5th of 10',
       '18th of 23', '14th of 19', '10th of 24', '10th of 13',
       '18th of 32', '17th of 25', '8th of 18', '14th of 17',
       '14th of 22', '20th of 22', '10th of 40', '20th of 21',
       '14th of 28', '14th of 24', '12th of 27', '2nd of 4', '24th of 25',
       '10th of 47', '15th of 22', '15th of 29', 'Gr of 28', '11th of 24',
       '18th of 27', '18th of 29', '12th of 12', '8th of 20',
       '11th of 18', '16th of 29', '23rd of 34', '11th of 28',
       '3rd of 28', '20th of 32', '21st of 28', '6th of 12', '10th of 35',
       '16th of 19', '10th of 10', '14th of 27', '16th of 20',
       '11th of 25', '10th of 23', '2nd of 6', 'Gr', '2nd 

### Additional Rooms & Additional Rooms.1

In [35]:
flat_data["Additional Rooms"].str.replace(r"\([a-z\s,]+\)","",regex=True).unique()

array([nan, '1 rooms', '3 rooms', '2 rooms'], dtype=object)

In [40]:
flat_data["Additional Rooms.1"].str.replace(r"\([a-z\s,]+\)","",regex=True).unique()

array([nan, '1 rooms', '3 rooms', '2 rooms'], dtype=object)

In [49]:
flat_data[["Additional Rooms.1","Additional Rooms"]][~(flat_data["Additional Rooms.1"].isnull())&(flat_data["Additional Rooms.1"].isnull())]

Empty DataFrame
Columns: [Additional Rooms.1, Additional Rooms]
Index: []

### Balconies & Balconies.1

In [36]:
flat_data["Balconies"].unique()

array([nan,  1.,  3.,  4.,  2.,  5.])

In [62]:
flat_data["Balconies"].unique()

array([nan,  1.,  3.,  4.,  2.,  5.])

In [63]:
flat_data[["Balconies.1","Balconies"]][~(flat_data["Balconies.1"].isnull())&(flat_data["Balconies.1"].isnull())]

Empty DataFrame
Columns: [Balconies.1, Balconies]
Index: []

### Security Deposit

In [37]:
flat_data["Security Deposit"].unique()

array(['No Deposit', nan], dtype=object)

### Facing

In [39]:
flat_data["Facing"].unique()

array(['North', 'East', nan, 'SouthEast', 'NorthEast', 'West',
       'NorthWest', 'South', 'SouthWest'], dtype=object)

### Ownership Type

In [85]:
flat_data["Ownership Type"].unique()

array([nan, 'Freehold', 'Leasehol', 'Co-Operative Society',
       'Power of Attorney'], dtype=object)

### Booking Amount

In [88]:
pd.to_numeric(flat_data["Booking Amount"].str.replace(",","")).unique()

array([      nan,    51000.,  9000000.,   140000.,   200000.,   300000.,
         800000.,   100000.,   600000.,  2000000.,   500000.,    50000.,
        1000000.,  3000000.,   190000.,   170000.,   650000.,   250000.,
        1100000.,   120000.,    89000.,    73500.,    93000.,    94000.,
          91000.,    86000.,    97000.,    68000.,    96000.,    82000.,
          71000.,    78000.,    61000.,    65000.,   156000.,    84000.,
          67000.,    98000.,    95000.,   230000.,   165000.,    85000.,
         360000.,    81000.,    86380.,    76000.,    20000.,   123800.,
        1700000.,  4080000.,  5380000.,  5299200.,  6000000.,   530000.,
         550000.,   900000.,  1200000.,  1600000.,  4460000.,  8860000.,
         762650.,   740000.,   404700., 10000000.,  5000000.,  6500000.,
        1500000.,    70000.,  4500000.,   130000.,  6400000.,  1040000.,
        1030400.,   700000.,    21000.,  2500000.,  1080000.,   960000.,
        4070000.,  6250000.,  3465000.,  2700000., 

### Overlooking

In [92]:
flat_data["Overlooking"].unique()

array([nan, 'Corner, Garden View, Pool View, Road View',
       'Garden View, Pool View, Road View', 'Road View',
       'Pool View, Garden View', 'Road View, Pool View, Garden View',
       'Garden View', 'Road View, Corner',
       'Pool View, Corner, Garden View', 'Road View, Pool View, Corner',
       'Road View, Pool View', 'Road View, Garden View',
       'Corner, Garden View, Road View', 'Corner, Garden View',
       'Corner, Garden View, Pool View', 'Corner', 'Corner, Road View',
       'Garden View, Pool View'], dtype=object)

### Amenities(1)

In [166]:
temp=flat_data["1"].replace(r"[]",np.nan).str.lower().str.replace(", ",",").str.replace(r"lift(s)","lift").str.replace(r"\[|\]","",regex=True).str.split(",",expand=True)

In [167]:
temp.shape

(9132, 24)

In [172]:
l

['sports facility',
 'shopping mall',
 'maintenance staff',
 'jogging track',
 'atm',
 'gymnasium',
 'indoor games',
 'rain water harvesting',
 'swimming pool',
 'intercom',
 'cafeteria',
 'full power backup',
 'lift',
 'childrens play area',
 'car parking',
 'landscaped gardens',
 '24 x 7 security',
 'club house',
 'staff quarter',
 'hospital',
 'multipurpose room',
 'vaastu compliant',
 'golf course',
 'school']

In [168]:
l=[]


for col in range(24):
    for row in range(9132):
        x=temp.iloc[row,col]
        if x not in l:
            l.append(x)

In [173]:
temp1=flat_data["1"].replace(r"[]",np.nan).str.lower()

In [128]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [175]:
flat_data.loc[3117,"1"]

'[Maintenance Staff, ATM, Intercom, Lift(s), Golf Course, Full Power Backup, Cafeteria, Childrens play area, Rain Water Harvesting, Club House, Staff Quarter, Swimming Pool, Multipurpose Room, Indoor Games, Car Parking, Hospital, School, Shopping Mall, 24 X 7 Security, Jogging Track, Sports Facility, Gymnasium, Vaastu Compliant]'

In [177]:
pd.DataFrame({e:temp1.str.contains(f"{e}") for e in l}).shape

(9132, 24)

In [176]:
l2=[]
temp1.to_frame().assign(**{e:temp1.str.contains(f"{e}") for e in l}).sample(5)
    

1 sports facility  \
8398  [indoor games, atm, jogging track, lift(s), in...            True   
6810  [gymnasium, lift(s), swimming pool, full power...           False   
8472  [gymnasium, lift(s), jogging track, indoor gam...           False   
3740  [jogging track, atm, maintenance staff, lift(s...            True   
341                                                 NaN             NaN   

     shopping mall maintenance staff jogging track    atm gymnasium  \
8398          True             False          True   True      True   
6810         False             False          True  False      True   
8472         False             False          True  False      True   
3740          True              True          True   True      True   
341            NaN               NaN           NaN    NaN       NaN   

     indoor games rain water harvesting swimming pool intercom cafeteria  \
8398         True                  True          True     True     False   
6810         True                  True          True    False     False   
8472         True                 False          True    False     False   
3740         True                  True          True     True      True   
341           NaN                   NaN           NaN      NaN       NaN   

     full power backup  lift childrens play area car parking  \
8398              True  True                True        True   
6810              True  True                True        True   
8472              True  True                True        True   
3740              True  True                True        True   
341                NaN   NaN                 NaN         NaN   

     landscaped gardens 24 x 7 security club house staff quarter hospital  \
8398               True            True       True          True     True   
6810              False            True       True         False    False   
8472              False            True       True         False    False   
3740              False            True       True          True     True   
341                 NaN             NaN        NaN           NaN      NaN   

     multipurpose room vaastu compliant golf course school  
8398              True             True       False   True  
6810             False            False       False  False  
8472             False            False       False  False  
3740              True             True       False   True  
341                NaN              NaN         NaN    NaN

### 2

In [184]:
flat_data["2"].unique()

array(['[Gas connection]', '[Gas connection, AC]', nan, '[AC]',
       '[BED, Dining Table, Microwave, Sofa, Refrigerator, TV, AC]',
       '[BED, Dining Table, Microwave, TV]', '[Dining Table, AC]',
       '[Washing Machine, BED, AC, Gas connection, Dining Table, Microwave, Sofa, Refrigerator, TV]',
       '[BED, Gas connection, Dining Table, AC]', '[BED, AC]',
       '[Gas connection, Microwave, AC]', '[Dining Table, TV, AC]',
       '[Dining Table, Refrigerator, AC]', '[TV, AC]',
       '[Dining Table, Washing Machine, AC]', '[Gas connection, TV, AC]',
       '[Gas connection, Microwave]', '[Dining Table]',
       '[BED, Gas connection, Dining Table, Microwave, Sofa, Refrigerator, TV, AC]',
       '[BED, Gas connection, Washing Machine]',
       '[BED, Dining Table, Washing Machine, TV]',
       '[Gas connection, Refrigerator, TV, AC]', '[Gas connection, Sofa]',
       '[Gas connection, TV]', '[Gas connection, Washing Machine]',
       '[Gas connection, Microwave, Refrigerator, TV, 

In [187]:
#flat_data[(flat_data["2"]!=r"[]")&~(flat_data["2"].isnull())][["Status.1","2"]]

**Observations:**
1. Name
    - contain multiple informations within the single attribute

2. Size
    - values contain in sq.ft. unit
    - the data type is inappropriate, it must be int

3. Address
    - There are total 724 unique values, can be reduce.

4. Price
    - values are contain in Cr./ L. unit need to be convert into number
    - some values contain unwanted suffix `Negotiable`
    - data type is inappropriate

5. Rate
    - values are contain in sq.ft. unit
    - values contain unwanted suffix
    - data type is inappropriate
    
6. carpet area
    - values are contain in sq.ft. unit
    - data type is inappropriate

7. Status1 & Status
    - `Property Status` name suggested for status attribute and `Furnishing Status` name suggested for status.1

8. Age of Property & Age of Property.1
    - inconsistent values
    - both attribute need to be combine

9. additional rooms & additional rooms.1
    - need to be combined

10. balconies & balconies.1
    - need to be combined

11. New/Resale & New/Resale.1
    - need to be combined
12. Amenities(1)
    - unnormalized form data
13. furnish_details(2)
    - unnormalized form data








 

In [204]:
def data_clean(df:pd.DataFrame)->pd.DataFrame:

    df_= (
        df
        .assign(   
            bedroom=df["name"].str.extract(r"(\d+)"),
            type1=df["name"].str.split(" ",n=1).str.get(1).str.extract(r"([A-Za-z]+)"),
            type2=df["name"].str.split(" ",n=1).str.get(1).str.replace(r"^([A-Za-z]+\s)","",regex=True),
            size=df["size"].str.replace(",","").str.extract(r"(\d+)"),
            address=(flat_data.address.str.lower().str.replace(r'[A-Za-z0-9\s]*?4 greater noida wes+[A-Za-z0-9\s]*',"4 greater noida west",regex=True)
                    .str.replace(r'[A-Za-z0-9\s]*?(extension|ext)[A-Za-z0-9\s]*',"extension, noida",regex=True)
                    .str.replace(r'[A-Za-z0-9\s]*?ambedkar city[A-Za-z0-9\s]*',"ambedkar city, noida",regex=True)
                    .str.replace(r'[A-Za-z0-9\s]*?yamuna express[A-Za-z0-9\s]*',"yamuna expressway, noida",regex=True)
                    .str.replace("surajpur site 4, noida","surajpur site 4, greater noida")
                    .str.replace("knowledge park 3, noida","knowledge park 3, greater noida")
                    .str.replace("sikandarpur village, noida","sector 87, noida")
                    .str.replace("dallupura, noida","sector 10, noida")
                    .str.replace("pristine avenuegaur city road, noida","gaur city, greater noida")
                    .str.replace("migsun vilaasaeta ii, noida","s block, greater noida")
                    .str.replace("nirala greenshiregreater noida west road, noida","greater noida west road, noida")
                    .str.replace("home and soul f premiereyeida, noida","jal vayu vihar, noida")
                    .str.replace("apte gra indraprastha phase 2shramik kunj, noida","sector 93, noida")
                    .str.replace("apte gra indraprastha phase 2shramik kunj, noida","sector 93, noida")
                    .str.replace(r'[0-9a-z ]+sector',"sector" ,regex=True).str.replace("-",'')
                    .str.replace("noida, ",'')   
                    .str.replace(r"sector[\s0]*", "sector ",regex=True)
                    .str.replace(r"[A-Za-z0-9\s,]*greater noida[A-Za-z0-9\s,]*","greater-noida",regex=True)
                    .str.replace(r"[,]* noida",", noida",regex=True)
                    .str.replace('dream house sr construction dream house sector 73, noida',"sector 73, noida")
                    .str.replace(r"sector 93[\sa-z]+, noida","sector 93, noida",regex=True)
                    .replace("sector 143b, noida","sector 143, noida")
                    .replace('sector 16b, noida','sector 16, noida')
                    .replace('sector 140a, noida','sector 140, noida')
                    .replace('sector 15a, noida','sector 15, noida')
                    .replace('sector 25a, noida','sector 25, noida')
                    .replace("greater-noida", "greater noida")
                    
                        ),
            price_in_lakh=((df["price"].str.replace("EMI","").str.extract(r"(Cr|L)").replace({"L":"1","Cr":"100"}).astype(float))*
                            (df["price"].str.replace("EMI","").str.extract(r"([0-9.]+)").astype(float))),
            rate= pd.to_numeric(  df["rate"].str.extract(r"([0-9,]+)").loc[:,0].str.replace(",","")),
            carpet_area= pd.to_numeric(  df["Carpet area"].str.extract(r"([0-9,]+)").loc[:,0].str.replace(",","")),
            age_of_property=(df["Age of Property"].fillna("")+df["Age of Property.1"].fillna("")).str.extract(r"([0-9]+ y)").loc[:,0].str.replace(" y",""),
            floor=pd.to_numeric(df["Floor"].str.replace("Gr","0").str.extract(r"(\d+)").iloc[:,0]),
            total_floor=pd.to_numeric(df["Floor"].str.extract(r"(\s\d+)").iloc[:,0]),
            additional_rooms=pd.to_numeric(((df["Additional Rooms"].fillna(""))+(df["Additional Rooms.1"].fillna(""))).str.replace(r"\([a-z\s,]+\)","",regex=True).str.extract(r"(\d+)").loc[:,0]),
            price_negotiable=(df["Price Negotiable"].fillna("")+df["Price Negotiable.1"].fillna("")),
            balconies=(df["Balconies"].fillna(0)+df["Balconies.1"].fillna(0)).replace(0,np.nan),
            type_of_sale=(df["New/Resale"].fillna(""))+(df["New/Resale.1"].fillna("")),
            booking_amount_in_lakh =(pd.to_numeric(df["Booking Amount"].str.replace(",",""))/100000)
            )
                      
    )


    temp1=df["1"].replace(r"[]",np.nan).str.lower().str.replace(", ",",").str.replace(r"lift(s)","lift").str.replace(r"\[|\]","",regex=True).str.split(",",expand=True)
    l=[]
    for col in range(24):
        for row in range(9132):
            x=temp1.iloc[row,col]
            if x not in l:
                l.append(x)
    l.remove(np.nan)
    l.remove(None)
    temp1=pd.DataFrame({f"amenities_{e}":df["1"].replace(r"[]",np.nan).str.lower().str.contains(f"{e}") for e in l})

    temp2=df["2"].replace(r"[]",np.nan).str.lower().str.replace(", ",",").str.replace(r"lift(s)","lift").str.replace(r"\[|\]","",regex=True).str.split(",",expand=True)
    l=[]
    for col in range(11):
        for row in range(9132):
            x=temp2.iloc[row,col]
            if x not in l:
                l.append(x)
    l.remove(np.nan)
    l.remove(None)
    temp2=pd.DataFrame({f"furnish_detail_{e}":df["2"].replace(r"[]",np.nan).str.lower().str.contains(f"{e}") for e in l})


    df_=(
        pd.concat([df_,temp1,temp2],axis=1)
        .rename(columns=str.lower)
        .drop(columns=["price","name","carpet area","age of property","age of property.1","additional rooms","additional rooms.1","price negotiable","price negotiable.1","balconies.1","new/resale",
                            "new/resale.1","link","name","1","2"]) 
    )
    df_.columns = [col.replace(' ', '_') for col in df_.columns]
    return df_

        

In [206]:
data_clean(flat_data).shape

(9132, 62)